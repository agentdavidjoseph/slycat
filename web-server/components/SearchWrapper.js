import React from "react";
import ProjectsList from "components/ProjectsList"
import { ModelsList } from 'components/ModelsList';

class SearchWrapper extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      initialItems: this.props.items,
      items: [],
      searchQuery: '',
    };
    this.filterList = this.filterList.bind(this);
    // console.log(this.props.items);
  }

  componentWillMount() {
      this.setState({items: this.state.initialItems});
  }

  filterList(e) {
    // console.log(`filter ${this}`);
    var updatedList = this.state.initialItems;
    var trimSearchQuery = e.target.value.trim();
    var lowerTrimSearchQuery = trimSearchQuery.toLowerCase();
    updatedList = updatedList.filter(function(item){
      // console.log('item: ' + item);
      return item.name.toLowerCase().search(lowerTrimSearchQuery) !== -1
             || item.description.toLowerCase().search(lowerTrimSearchQuery) !== -1
             // Why is this line here, it's identical to the previous one? Alex commenting out for now.
             // || item.description.toLowerCase().search(e.target.value.toLowerCase()) !== -1 
             ;
    });
    this.setState({items: updatedList, searchQuery: trimSearchQuery});
  };

  render() {

    let searchField = 
      <input className="form-control mb-3" 
        style={{width: '13rem'}}
        type="search" 
        placeholder={`Filter ${this.props.type}`} 
        aria-label={`Filter ${this.props.type}`} 
        onChange={this.filterList}
      />
    ;

    let message = '';

    // When there are no items returned but there are initial items
    if(this.props.items.length > 0 && this.state.items.length == 0)
    {
      message = 
        <div className="container">
          <div className="alert alert-warning" role="alert">
            <p>No {this.props.type} match the current search - <strong>{this.state.searchQuery}</strong></p>
            <p className="mb-0">Clear it to see all {this.props.type}, or change it to search again.</p>
          </div>
        </div>
      ;
    }
    // When there are no models (i.e., empty project)
    else if (this.props.items.length == 0 && this.props.type == 'models')
    {
      message = 
        <div className="container">
          <div className="alert alert-warning" role="alert">
              <p>There are no models in this project.</p>
              <p className="mb-0">You can add a model by using the Create menu above.</p>
          </div>
        </div>
      ;
      searchField = null;
    }
    // When there are no projects (i.e., empty site)
    else if (this.props.items.length == 0 && this.props.type == 'projects')
    {
      message = 
        <div className="container">
          <div className="alert alert-warning" role="alert">
              <p>There are no projects.</p>
              <p className="mb-0">You can add a project by using the Create menu above.</p>
          </div>
        </div>
      ;
      searchField = null;
    }

    let list = <ProjectsList projects={this.state.items} />
    if(this.props.type == 'models')
    {
      list = <ModelsList models={this.state.items} />
    }

    return (
      <React.Fragment>
        <div className="container pb-0">
          <div className="d-flex justify-content-between">
            <h3 className="px-4 text-capitalize">{this.props.type}</h3>
            {searchField}
          </div>
        </div>
        {list}
        {message}
      </React.Fragment>
    );
  }
}

export default SearchWrapper;