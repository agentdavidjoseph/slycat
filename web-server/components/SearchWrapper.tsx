import React from "react";
import ProjectsList from "components/ProjectsList"
import { ModelsList } from 'components/ModelsList';

/**
 * @param {items} list of item objects
 * @param {type} string type
 */
export interface SearchWrapperProps { 
  items: Item[]
  type: string
}

/**
 * @param {initialItems} list of Item objects
 * @param {items} list of Item objects
 * @param {searchQuery} string to regex on
 */
export interface SearchWrapperState {
  initialItems: Item[],
  items: Item[],
  searchQuery: string,
}
/**
 * @param {name} string name
 * @param {description} string description
 * @param {cretor} name of the creator as string
 * @param {created} string representation of date created
 * @interface Item
 */
interface Item {
  name: string
  description: string
  creator: string
  created: string
}

/**
 * class that filters on item objects and search term
 */
export default class SearchWrapper extends React.Component<SearchWrapperProps, SearchWrapperState> {

  public constructor(props: SearchWrapperProps) {
    super(props);
    this.state = {
      initialItems: this.props.items,
      items: [],
      searchQuery: '',
    };
  }

  componentWillMount() {
      this.setState({items: this.state.initialItems});
  }

  /**
   * match two strings
   * @param {one} string 1
   * @param {two} strin twop
   * @returns boolean true if they match
   * @memberof SearchWrapper
   */
  matchStrings = (stringOne: string, stringTwo: string): boolean =>{
    return stringOne.toLowerCase().search(stringTwo.toLowerCase()) !== -1
  }

  /**
   * @param {trimSearchQuery} string that has been trimmed to use as a matcher
   *
   * @memberof SearchWrapper
   */
  filterList = (trimSearchQuery: string): void => {
    const updatedList = this.state.initialItems.filter(
      ({name, description, creator, created}) => {
      return this.matchStrings(name,trimSearchQuery)
             || this.matchStrings(description,trimSearchQuery)
             || this.matchStrings(creator, trimSearchQuery)
             || this.matchStrings(created, trimSearchQuery)
             ;
    });
    this.setState({items: updatedList, searchQuery: trimSearchQuery});
  };

  /**
   * creates the search input field
   *
   * @memberof SearchWrapper
   */
  getSearchField = () : JSX.Element | null => {
    if ((this.props.items.length == 0 && this.props.type == 'models')||
    (this.props.items.length == 0 && this.props.type == 'models')) {
      return null;
    };
    return <input className="form-control mb-3" 
              style={{width: '13rem'}}
              type="search" 
              placeholder={`Filter ${this.props.type}`} 
              aria-label={`Filter ${this.props.type}`} 
              onChange={(e)=>this.filterList(e.target.value.trim().toLowerCase())}
            />;
  }

  /**
   * populate the model or projects list depending on the type passed to props
   *
   * @memberof SearchWrapper
   */
  getList = () : JSX.Element => {
    return this.props.type == 'models'? 
    <ModelsList models={this.state.items} />:
    <ProjectsList projects={this.state.items} />;
  }

  /**
   * formats the message display to display when the list are empty
   *
   * @memberof SearchWrapper
   */
  getMessage = () : JSX.Element | null => {
    let message: JSX.Element[]| null = null;
    // When there are no items returned but there are initial items
    if(this.props.items.length > 0 && this.state.items.length == 0)
    {
      message = [
            <p key={1}>No {this.props.type} match the current search - <strong>{this.state.searchQuery}</strong></p>,
            <p key={2} className="mb-0">Clear it to see all {this.props.type}, or change it to search again.</p>
      ];
    }
    // When there are no models (i.e., empty project)
    else if (this.props.items.length == 0 && this.props.type == 'models')
    {
      message = [
              <p key={1}>There are no models in this project.</p>,
              <p key={2} className="mb-0">You can add a model by using the Create menu above.</p>
      ];
    }
    // When there are no projects (i.e., empty site)
    else if (this.props.items.length == 0 && this.props.type == 'projects')
    {
      message = [
        <p key={1}>There are no projects.</p>,
        <p key={2} className="mb-0">You can add a project by using the Create menu above.</p>
      ];
    }
    return  message ? (
        <div className="container">
          <div className="alert alert-warning" role="alert">
            {message}
          </div>
        </div>
        ):null;
  }
  
  render() {
    return (
      <div>
        <div className="container pb-0">
          <div className="d-flex justify-content-between">
            <h3 className="px-4 text-capitalize">{this.props.type}</h3>
            {this.getSearchField()}
          </div>
        </div>
        {this.getList()}
        {this.getMessage()}
      </div>
    );
  }
}
