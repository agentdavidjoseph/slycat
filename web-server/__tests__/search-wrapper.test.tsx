
import React from 'react';
import $ from 'jquery';
// TODO the html loader breaks here
import SearchWrapper, {SearchWrapperProps, SearchWrapperState} from "components/SearchWrapper";
import { mount } from "enzyme";
(global as any).$ = (global as any).jQuery = $;
describe('when loading a slycat selector',() =>{

  const properties: SearchWrapperProps = {
    items:[{
      name:'test1', 
      description:'test1',
      created:'test1',
      creator: 'test1'
    }],
    type:'models'
  };
  const render = mount(
    <SearchWrapper
      {...properties}
    />
  );
  //TODO: fix jest to use html loader
  test('we have expected props on initial load', () => {
    const props = render.props() as SearchWrapperProps;
    expect(props).toMatchSnapshot();
  });

  test('we see no state', () => {
    const state = render.state() as SearchWrapperState;
    expect(state).toMatchSnapshot();
  });

  test('we should have a full initialization', () => {
    expect(render).toMatchSnapshot();
  });

  test('we should onChange function get value', () => {
    jest.spyOn(render.instance(), 'filterList');
    jest.spyOn(render.instance(), 'matchStrings');
    render.find('input').simulate('change',{ target: { value: 'test' } });
    expect(render.instance().filterList).toHaveBeenCalledWith('test');
    expect(render.instance().matchStrings).toHaveBeenCalledWith('test1','test');
  });

  test('we should have null input if there is no list', () => {
    const emptyProperties: SearchWrapperProps = {
      items:[],
      type:'models'
    };
    const emptyRender = mount(
      <SearchWrapper
        {...emptyProperties}
      />
    );
    expect(emptyRender).toMatchSnapshot();
  });

});

export default undefined