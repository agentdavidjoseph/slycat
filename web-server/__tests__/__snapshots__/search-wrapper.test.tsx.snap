// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when loading a slycat selector we have expected props on initial load 1`] = `
Object {
  "items": Array [
    Object {
      "created": "test1",
      "creator": "test1",
      "description": "test1",
      "name": "test1",
    },
  ],
  "type": "models",
}
`;

exports[`when loading a slycat selector we see no state 1`] = `
Object {
  "initialItems": Array [
    Object {
      "created": "test1",
      "creator": "test1",
      "description": "test1",
      "name": "test1",
    },
  ],
  "items": Array [
    Object {
      "created": "test1",
      "creator": "test1",
      "description": "test1",
      "name": "test1",
    },
  ],
  "searchQuery": "",
}
`;

exports[`when loading a slycat selector we should have a full initialization 1`] = `
<SearchWrapper
  items={
    Array [
      Object {
        "created": "test1",
        "creator": "test1",
        "description": "test1",
        "name": "test1",
      },
    ]
  }
  type="models"
>
  <div>
    <div
      className="container pb-0"
    >
      <div
        className="d-flex justify-content-between"
      >
        <h3
          className="px-4 text-capitalize"
        >
          models
        </h3>
        <input
          aria-label="Filter models"
          className="form-control mb-3"
          onChange={[Function]}
          placeholder="Filter models"
          style={
            Object {
              "width": "13rem",
            }
          }
          type="search"
        />
      </div>
    </div>
    <ModelsList
      models={
        Array [
          Object {
            "created": "test1",
            "creator": "test1",
            "description": "test1",
            "name": "test1",
          },
        ]
      }
    >
      <Spinner>
        <div
          className="d-flex justify-content-center mt-4"
        >
          <div
            className="spinner-border"
            role="status"
          >
            <span
              className="sr-only"
            >
              Loading...
            </span>
          </div>
        </div>
      </Spinner>
    </ModelsList>
  </div>
</SearchWrapper>
`;

exports[`when loading a slycat selector we should have null input if there is no list 1`] = `
<SearchWrapper
  items={Array []}
  type="models"
>
  <div>
    <div
      className="container pb-0"
    >
      <div
        className="d-flex justify-content-between"
      >
        <h3
          className="px-4 text-capitalize"
        >
          models
        </h3>
      </div>
    </div>
    <ModelsList
      models={Array []}
    >
      <Spinner>
        <div
          className="d-flex justify-content-center mt-4"
        >
          <div
            className="spinner-border"
            role="status"
          >
            <span
              className="sr-only"
            >
              Loading...
            </span>
          </div>
        </div>
      </Spinner>
    </ModelsList>
    <div
      className="container"
    >
      <div
        className="alert alert-warning"
        role="alert"
      >
        <p
          key="1"
        >
          There are no models in this project.
        </p>
        <p
          className="mb-0"
          key="2"
        >
          You can add a model by using the Create menu above.
        </p>
      </div>
    </div>
  </div>
</SearchWrapper>
`;
